File: day1/solution part 2.py
l1=[]
l2=[]

sum=0

# two lists

with open("data.txt", 'r') as file:
    for line in file:
        n1, n2 = map(int, line.split())
        l1.append(n1)
        l2.append(n2)

counts = {}
sum = 0

for num in l2:
    counts[num] = counts.get(num, 0) + 1

for num in l1:
    sum += num * counts.get(num, 0)

print(sum)

File: day1/solution part 2.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <fstream>
#include <sstream>
#include <string>

int main() {
    std::ifstream file("data.txt");
    std::string line, num;

    std::vector<int> v1, v2;
    while (std::getline(file, line)) {
        std::stringstream ss(line);
        ss >> num;
        v1.push_back(stoi(num));
        ss >> num;
        v2.push_back(stoi(num));
    }

    std::unordered_map<int, int> counts;  // use a hashmap for O(N) complexity
    for (int num : v2) {
        counts[num]++;
    }

    int sum = 0;
    for (int num : v1) {
        sum += num * counts[num];
    }

    std::cout << sum << std::endl;

    return 0;
}
